crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -0,0 +1,1285 @@
+                if let Some(qarg_sample) = qarg_sample {
+                    if qarg_len != *size_dict.entry(qarg_sample.len()).or_insert(0) {
+                        incomplete_basis_gates.push(inst.to_owned());
+                    }
+                }
+            }
+        }
+        if strict_direction {
+            self.non_global_strict_basis = Some(incomplete_basis_gates);
+            self.non_global_strict_basis.as_ref().unwrap()
+        } else {
+            self.non_global_basis = Some(incomplete_basis_gates.to_owned());
+            self.non_global_basis.as_ref().unwrap()
+        }
+    }
+
+    /// Get all non_global operation names.
+    pub fn get_non_global_op_names(&mut self, strict_direction: bool) -> Option<&[String]> {
+        if strict_direction {
+            if self.non_global_strict_basis.is_some() {
+                return self.non_global_strict_basis.as_deref();
+            }
+        } else if self.non_global_basis.is_some() {
+            return self.non_global_basis.as_deref();
+        }
+        return Some(self.generate_non_global_op_names(strict_direction));
+    }
+
+    /// Gets all the operation names that use these qargs. Rust native equivalent of ``BaseTarget.operation_names_for_qargs()``
+    pub fn operation_names_for_qargs(
+        &self,
+        qargs: &Option<Qargs>,
+    ) -> Result<HashSet<&String, RandomState>, TargetKeyError> {
+        // When num_qubits == 0 we return globally defined operators
+        let mut res: HashSet<&String, RandomState> = HashSet::default();
+        let mut qargs = qargs;
+        if self.num_qubits.unwrap_or_default() == 0 || self.num_qubits.is_none() {
+            qargs = &None;
+        }
+        if let Some(qargs) = qargs.as_ref() {
+            if qargs
+                .iter()
+                .any(|x| !(0..self.num_qubits.unwrap_or_default()).contains(&x.index()))
+            {
+                return Err(TargetKeyError::new_err(format!(
+                    "{:?} not in Target",
+                    qargs
+                )));
+            }
+        }
+        if let Some(Some(qarg_gate_map_arg)) = self.qarg_gate_map.get(qargs).as_ref() {
+            res.extend(qarg_gate_map_arg);
+        }
+        for name in self._gate_name_map.keys() {
+            if self.variable_class_operations.contains(name) {
+                res.insert(name);
+            }
+        }
+        if let Some(qargs) = qargs.as_ref() {
+            if let Some(global_gates) = self.global_operations.get(&(qargs.len() as u32)) {
+                res.extend(global_gates)
+            }
+        }
+        if res.is_empty() {
+            return Err(TargetKeyError::new_err(format!(
+                "{:?} not in target",
+                qargs
+            )));
+        }
+        Ok(res)
+    }
+
+    /// Returns rust-native operation instances present in the Target that affect the provided qargs.
+    pub fn ops_from_qargs(
+        &self,
+        qargs: &Option<Qargs>,
```

Comment:
It's a bit more idiomatic to use `Option<&Qargs>` because you can convert from `&Option<T>` -> `Option<&T>` with `Option.as_ref()` but you can't convert the other way.

Reply:
This will be a bit tricky to implement as it will require us to perform cloning inside of the function. I will leave this item open for further discussion.

Reply:
I implemented an alternative structure that wraps around the `IndexMap` and can process `Option<&K>` as an acceptable index.